File: scripts/codex/render-bridge-mcp-catalogue.py
#!/usr/bin/env python3
"""Render the MCP catalogue the provider bridges read.

`.rulesync/mcp.jsonc` is the only MCP source. The installer generates its Codex
projection with Rulesync; this renderer keeps only each server's launch keys
(`command` and `args`, or `url`) so the Claude and Copilot bridges can give a
role exactly its contract's servers without parsing TOML or restating any
server definition.
"""

from __future__ import annotations

import argparse
import json
import os
import tempfile
import tomllib
from pathlib import Path

LAUNCH_KEYS = ("command", "args", "url")


def render(mcp_source: Path) -> str:
    try:
        servers = tomllib.loads(mcp_source.read_text(encoding="utf-8")).get("mcp_servers")
    except (OSError, tomllib.TOMLDecodeError) as error:
        raise RuntimeError(f"unable to read MCP source {mcp_source}: {error}") from error
    if not isinstance(servers, dict) or not servers:
        raise RuntimeError(f"MCP source declares no mcp_servers: {mcp_source}")
    catalogue = {
        name: {key: server[key] for key in LAUNCH_KEYS if key in server}
        for name, server in sorted(servers.items())
        if isinstance(server, dict)
    }
    return json.dumps(catalogue, indent=2, sort_keys=True) + "\n"


def write(output: Path, content: str) -> None:
    output.parent.mkdir(parents=True, exist_ok=True)
    descriptor, temporary = tempfile.mkstemp(prefix=f".{output.name}.", suffix=".tmp", dir=output.parent)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8") as stream:
            stream.write(content)
        os.chmod(temporary, 0o644)
        os.replace(temporary, output)
    finally:
        Path(temporary).unlink(missing_ok=True)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--mcp-source", type=Path, required=True, help="Codex config.toml Rulesync generated from .rulesync/mcp.jsonc.")
    parser.add_argument("--output", type=Path, required=True, help="Catalogue path, normally $CODEX_HOME/provider-runtime/mcp-servers.json.")
    parser.add_argument("--check", action="store_true", help="Report drift without writing.")
    args = parser.parse_args()
    try:
        rendered = render(args.mcp_source)
    except RuntimeError as error:
        parser.error(str(error))
    if args.check:
        if args.output.is_file() and not args.output.is_symlink() and args.output.read_text(encoding="utf-8") == rendered:
            print(f"ok bridge MCP catalogue {args.output}")
            return 0
        print(f"missing-or-drifted bridge MCP catalogue {args.output}")
        return 1
    write(args.output, rendered)
    print(f"rendered bridge MCP catalogue into {args.output}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/config/render-bridge-mcp-catalogue.ts
import { existsSync, lstatSync, readFileSync } from "node:fs";
import { ConfigError, atomicWrite, parseArgs, parseTomlFile, requiredArg, type TomlTable } from "./toml.ts";

const LAUNCH_KEYS = ["command", "args", "url"] as const;

export function renderBridgeMcpCatalogue(source: string): string {
  const servers = parseTomlFile(source, "MCP source").mcp_servers;
  if (!servers || typeof servers !== "object" || Array.isArray(servers) || Object.keys(servers).length === 0) {
    throw new ConfigError(`MCP source declares no mcp_servers: ${source}`);
  }
  const catalogue: Record<string, TomlTable> = {};
  for (const name of Object.keys(servers as TomlTable).sort()) {
    const server = (servers as TomlTable)[name];
    if (!server || typeof server !== "object" || Array.isArray(server)) continue;
    const launch: TomlTable = {};
    const serverTable = server as TomlTable;
    for (const key of LAUNCH_KEYS) if (serverTable[key] !== undefined) launch[key] = serverTable[key]!;
    catalogue[name] = launch;
  }
  return `${JSON.stringify(catalogue, null, 2)}\n`;
}

export function runBridgeMcpCatalogue(source: string, output: string, check = false): number {
  const rendered = renderBridgeMcpCatalogue(source);
  if (check) {
    if (existsSync(output) && !lstatSync(output).isSymbolicLink() && readFileSync(output, "utf8") === rendered) {
      console.log(`ok bridge MCP catalogue ${output}`); return 0;
    }
    console.log(`missing-or-drifted bridge MCP catalogue ${output}`); return 1;
  }
  atomicWrite(output, rendered);
  console.log(`rendered bridge MCP catalogue into ${output}`); return 0;
}

if (process.argv[1] === new URL(import.meta.url).pathname) {
  try { const { values, flags } = parseArgs(process.argv.slice(2)); process.exitCode = runBridgeMcpCatalogue(requiredArg(values, "mcp-source"), requiredArg(values, "output"), flags.has("check")); }
  catch (error) { console.error(`render-bridge-mcp-catalogue: ${error instanceof Error ? error.message : error}`); process.exitCode = 2; }
}
